src/config/builder.rs

```diff
@@ -0,0 +1,150 @@
+use super::{
+    compiler, default_data_dir, Config, GlobalOptions, SinkConfig, SinkOuter, SourceConfig,
+    TestDefinition, TransformConfig, TransformOuter,
+};
+use crate::event;
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+#[derive(Deserialize, Serialize, Debug, Default)]
+#[serde(deny_unknown_fields)]
+pub struct ConfigBuilder {
+    #[serde(flatten)]
+    pub global: GlobalOptions,
+    #[serde(default)]
+    pub sources: IndexMap<String, Box<dyn SourceConfig>>,
+    #[serde(default)]
+    pub sinks: IndexMap<String, SinkOuter>,
+    #[serde(default)]
+    pub transforms: IndexMap<String, TransformOuter>,
+    #[serde(default)]
+    pub tests: Vec<TestDefinition>,
+}
+
+impl Clone for ConfigBuilder {
+    fn clone(&self) -> Self {
+        // This is a hack around the issue of cloning
+        // trait objects. So instead to clone the config
+        // we first serialize it into JSON, then back from
+        // JSON. Originally we used TOML here but TOML does not
+        // support serializing `None`.
+        let json = serde_json::to_vec(self).unwrap();
+        serde_json::from_slice(&json[..]).unwrap()
+    }
+}
+
+impl ConfigBuilder {
+    pub fn build(self) -> Result<Config, Vec<String>> {
+        compiler::compile(self)
+    }
+
+    pub fn add_source<S: SourceConfig + 'static>(&mut self, name: &str, source: S) {
```

Comment:
API guidelines  suggest we allow the caller to determine copy and placement: https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control

How about `name: impl Into<String>`?

Reply:
This was all just copied but sure :smile: